plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -406,20 +406,23 @@ class producer_plugin_impl : public std::enable_shared_from_this<producer_plugin
 
       incoming_transaction_queue _pending_incoming_transactions;
 
-      void on_incoming_transaction_async(const transaction_metadata_ptr& trx, bool persist_until_expired, next_function<transaction_trace_ptr> next) {
+      void on_incoming_transaction_async(const packed_transaction_ptr& trx, bool persist_until_expired, next_function<transaction_trace_ptr> next) {
          chain::controller& chain = chain_plug->chain();
          const auto max_trx_time_ms = _max_transaction_time_ms.load();
          fc::microseconds max_trx_cpu_usage = max_trx_time_ms < 0 ? fc::microseconds::maximum() : fc::milliseconds( max_trx_time_ms );
 
-         auto after_sig_recovery =
-               [self = this, trx, persist_until_expired, next{std::move(next)}]() mutable {
-                  app().post(priority::low, [self, trx{std::move(trx)}, persist_until_expired, next{std::move(next)}]() {
-                     self->process_incoming_transaction_async( trx, persist_until_expired, next );
-                  });
-               };
-
-         transaction_metadata::start_recover_keys( trx, chain.get_thread_pool(), chain.get_chain_id(),
-                  max_trx_cpu_usage, std::move( after_sig_recovery ) );
+         auto future = transaction_metadata::start_recover_keys( trx, _thread_pool->get_executor(),
+                chain.get_chain_id(), fc::microseconds( max_trx_cpu_usage ), chain.configured_subjective_signature_length_limit() );
+         boost::asio::post( _thread_pool->get_executor(), [self = this, future{std::move(future)}, persist_until_expired, next{std::move(next)}]() mutable {
+            if( future.valid() ) {
+               future.wait();
+               app().post( priority::low, [self, future{std::move(future)}, persist_until_expired, next{std::move( next )}]() mutable {
+                  try {
+                     self->process_incoming_transaction_async( future.get(), persist_until_expired, std::move( next ) );
+                  } CATCH_AND_CALL(next);
+               } );
+            }
+         });
       }
```

Comment:
This is the key performance change. As odd as this looks, it is demonstratively better than before. And it is not the switch back to producer thread pool. Either thread pool (chain or producer) showed same results. The difference is in passing in a next `std::function` over calling `future.wait()`.